{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Projet UE MU4BM748\n",
    "\n",
    "Antoine Auvergne <Br>\n",
    "n° étudiant : 3522049"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Importation des bibliothèques"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from tkinter import *\n",
    "import numpy as np\n",
    "import os\n",
    "import re\n",
    "import copy\n",
    "from matplotlib.backends.backend_tkagg import (FigureCanvasTkAgg, NavigationToolbar2Tk)\n",
    "from matplotlib.backend_bases import key_press_handler\n",
    "from matplotlib.figure import Figure"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Extraction des données\n",
    "##### Protéomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "def extract_proteomes():\n",
    "    os.chdir('./Proteomes')\n",
    "    fichiers=!dir\n",
    "    noms=[]\n",
    "    \n",
    "    for Fi in fichiers:\n",
    "        if Fi[-4:]=='.faa' or Fi[-5:-1]=='.fasta':\n",
    "            fichier=Fi.split(' ')[-1]\n",
    "            noms.append('./Proteomes/'+fichier)\n",
    "    os.chdir('..')\n",
    "    data_proteomes={}\n",
    "    for fichier in noms:\n",
    "        motif=re.compile('[A-Za-z]{3}_[0-9]{9}[.]?[0-9]{0,2}')\n",
    "        cle=motif.search(fichier).group()\n",
    "        data_proteomes[cle]=[]\n",
    "        lignes=open(fichier,'r')\n",
    "        seq=''\n",
    "        nomprot=''\n",
    "        for ligne in lignes:\n",
    "            if ligne[0]=='>':\n",
    "                if seq!='' and nomprot!='':\n",
    "                    data_proteomes[cle].append([nomprot,len(seq)*3])\n",
    "                seq=''\n",
    "                nomprot=ligne.split(' ')[0][5:]\n",
    "#                location=(ligne.split('location=')[1]).split(']')[0].split('..')\n",
    "#                if location[0][0:4]=='comp':\n",
    "#                    location[0]=location[0].split('(')[1]\n",
    "#                    location[1]=location[1].split(')')[0]\n",
    "            else:\n",
    "                seq+=ligne.split('\\n')[0]\n",
    "        data_proteomes[cle].append([nomprot,len(seq)*3])\n",
    "        lignes.close()\n",
    "    return(data_proteomes)\n",
    "#data_proteomes=extract_proteomes()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### CD search"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "def extract_cog_fun():\n",
    "    fichier=open('cognames2003-2014.tab.txt','r',encoding='windows-1252')\n",
    "    dico_cog={}\n",
    "    for ligne in fichier.readlines():\n",
    "        if ligne[0]!='#' and ligne[0]!='':\n",
    "            dico_cog[ligne.split('\\t')[0]]=(ligne.split('\\t')[1],ligne.split('\\t')[2].split('\\n')[0])\n",
    "    fichier=open('fun2003-2014.tab.txt','r',encoding='windows-1252')\n",
    "    dico_fun={}\n",
    "    for ligne in fichier:\n",
    "        if ligne[0]!='#' and ligne[0]!='':\n",
    "            dico_fun[ligne.split('\\t')[0]]=ligne.split('\\t')[1].split('\\n')[0]\n",
    "    return(dico_cog,dico_fun)\n",
    "\n",
    "def extract_CDsearch():\n",
    "    os.chdir('./CD-search')\n",
    "    fichiers=!dir\n",
    "    noms=[]\n",
    "    for Fi in fichiers:\n",
    "        if Fi[-4:]=='.out':\n",
    "            fichier=Fi.split(' ')[-1]\n",
    "            noms.append('./CD-search/'+fichier)\n",
    "    os.chdir('..')\n",
    "    data_cd={}\n",
    "    for fichier in noms:\n",
    "        motif=re.compile('[A-Za-z]{3}_[0-9]{9}[.]?[0-9]{0,2}')\n",
    "        cle=motif.search(fichier).group()\n",
    "        data_cd[cle]={}\n",
    "        lignes=open(fichier,'r')\n",
    "        for ligne in lignes:\n",
    "            if ligne[0:2]=='Q#' and ligne.split('\\t')[7][0:3]=='COG':\n",
    "                cle2=ligne.split(' ')[2][5:]\n",
    "                ligne=ligne.split('\\t')\n",
    "                liste=[ligne[i] for i in [3,4,5,6,7,2]]\n",
    "                data_cd[cle][cle2]=liste\n",
    "        lignes.close()\n",
    "    return(data_cd)\n",
    "                \n",
    "                \n",
    "#dico_cog,dico_fun=extract_cog_fun()\n",
    "#data_cd=extract_CDsearch()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Blastp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def extract_blastp():\n",
    "    os.chdir('./Blastp')\n",
    "    fichiers=!dir\n",
    "    noms=[]\n",
    "    for Fi in fichiers:\n",
    "        if Fi[-4:]=='.out':\n",
    "            fichier=Fi.split(' ')[-1]\n",
    "            noms.append('./Blastp/'+fichier)\n",
    "    os.chdir('..')\n",
    "    data_blastp={}\n",
    "    for fichier in noms:\n",
    "        motif=re.compile('[A-Za-z]{3}_[0-9]{9}[.]?[0-9]{0,2}')\n",
    "        cle=motif.findall(fichier)\n",
    "        cle=cle[0],cle[1]\n",
    "        lignes=open(fichier,'r')\n",
    "        data_blastp[cle]={}\n",
    "        for ligne in lignes:\n",
    "            if ligne[0]!='#':\n",
    "                ligne=ligne.split('\\t')\n",
    "                nomprots=(ligne[0][4:],ligne[1])\n",
    "                data_blastp[cle][nomprots]=ligne[2:-1]\n",
    "          \n",
    "        \n",
    "        lignes.close()\n",
    "    return(data_blastp)\n",
    "#data_blastp=extract_blastp()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Prokaryotes complete genomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "def extract_compgenomes():\n",
    "    fichier=open('prokaryotes_complete-genomes.csv','r')\n",
    "    dico_genomes={}\n",
    "    for ligne in fichier:\n",
    "        if ligne[0]!='#':\n",
    "            ligne=ligne.split(',')\n",
    "            cle=ligne[5][1:-1]\n",
    "            dico_genomes[cle]=[ligne[0][1:-1],ligne[1][1:-1],ligne[2][1:-1],ligne[7],ligne[8],ligne[12],ligne[14][1:-1],ligne[15][1:-2]]\n",
    "    return(dico_genomes)\n",
    "\n",
    "#compgenomes=extract_compgenomes()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Récapitulatif des dictionnaires :\n",
    "##### compgenomes : dictionnaire [ code GCA ] <Br>\n",
    "    - indice 0 : Nom\n",
    "    - indice 1 : Groupe Monophylétique\n",
    "    - indice 2 : Espèce\n",
    "    - indice 3 : Longueur du génome (en Mb)\n",
    "    - indice 4 : %gc\n",
    "    - indice 5 : CDS\n",
    "    - indice 6 : FTP sur Genbank\n",
    "    - indice 7 : FTPS sur RefSeq\n",
    "\n",
    "##### dico_cog : dictionnaire [ code COG ]\n",
    "    - indice 0 : Code à 1 lettre de la fonction\n",
    "    - indice 1 : Nom\n",
    "    \n",
    "##### dico_fun : dictionnaire [ code à une lettre de la fonction ]\n",
    "    - renvoie la fonction associée à ce code\n",
    "    \n",
    "##### data_proteomes : dictionnaire [ code GCA ]\n",
    "liste de k éléments (k étant le nombre de protéines pour le génome). Dans chaque élément, on a :\n",
    "    - indice 0 : nom de la protéine\n",
    "    - indice 1 : longueur de la protéine\n",
    "   \n",
    "##### data_cd :  dictionnaire [ code GCA ] [ nom protéine ]\n",
    "    - indice 0 : début du match sur la séquence protéique\n",
    "    - indice 1 : fin du match sur la séquence protéique\n",
    "    - indice 2 : e-value associée au match\n",
    "    - indice 3 : bitscore\n",
    "    - indice 4 : code COG\n",
    "    - indice 5 : PSSM-ID\n",
    "\n",
    "##### data_blastp : dictionnaire [code GCA 1, code GCA 2] [nom protéine1, nom protéine2]\n",
    "    - indice 0 : pourcentage d'identité\n",
    "    - indice 1 : longueur du match\n",
    "    - indice 2 : nombre de mismatch\n",
    "    - indice 3 : nombre de gapopen\n",
    "    - indice 4 : début du match sur la query\n",
    "    - indice 5 : fin du match sur la query\n",
    "    - indice 6 : début du match sur la séquence comparée\n",
    "    - indice 7 : fin du match sur la séquence comparée\n",
    "    - indice 8 : e-value associée au match"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Interface graphique"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "ename": "UnicodeDecodeError",
     "evalue": "'utf-8' codec can't decode byte 0x91 in position 6014: invalid start byte",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mUnicodeDecodeError\u001b[0m                        Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-12-2793ce593f4f>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m    400\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    401\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 402\u001b[0;31m \u001b[0ma\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0minterface\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    403\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    404\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-12-2793ce593f4f>\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m      4\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcd\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mextract_CDsearch\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mblastp\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mextract_blastp\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 6\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcognames\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfonctioncog\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mextract_cog_fun\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      7\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgenomes\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mextract_compgenomes\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      8\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-11-7e9829bcaebf>\u001b[0m in \u001b[0;36mextract_cog_fun\u001b[0;34m()\u001b[0m\n\u001b[1;32m      2\u001b[0m     \u001b[0mfichier\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'cognames2003-2014.tab.txt'\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m'r'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m     \u001b[0mdico_cog\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m{\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 4\u001b[0;31m     \u001b[0;32mfor\u001b[0m \u001b[0mligne\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mfichier\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreadlines\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      5\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mligne\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m!=\u001b[0m\u001b[0;34m'#'\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0mligne\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m!=\u001b[0m\u001b[0;34m''\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      6\u001b[0m             \u001b[0mdico_cog\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mligne\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msplit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'\\t'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mligne\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msplit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'\\t'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mligne\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msplit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'\\t'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m2\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msplit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'\\n'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/lib/python3.7/codecs.py\u001b[0m in \u001b[0;36mdecode\u001b[0;34m(self, input, final)\u001b[0m\n\u001b[1;32m    320\u001b[0m         \u001b[0;31m# decode input (taking the buffer into account)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    321\u001b[0m         \u001b[0mdata\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mbuffer\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0minput\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 322\u001b[0;31m         \u001b[0;34m(\u001b[0m\u001b[0mresult\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mconsumed\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_buffer_decode\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0merrors\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfinal\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    323\u001b[0m         \u001b[0;31m# keep undecoded input until the next call\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    324\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mbuffer\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdata\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mconsumed\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mUnicodeDecodeError\u001b[0m: 'utf-8' codec can't decode byte 0x91 in position 6014: invalid start byte"
     ]
    }
   ],
   "source": [
    "class interface:\n",
    "    def __init__(self):\n",
    "        self.proteomes=extract_proteomes()\n",
    "        self.cd=extract_CDsearch()\n",
    "        self.blastp=extract_blastp()\n",
    "        self.cognames,self.fonctioncog=extract_cog_fun()\n",
    "        self.genomes=extract_compgenomes()\n",
    "        \n",
    "        # Création de la liste pour le menu déroulant\n",
    "        self.listemenu=[]\n",
    "        self.listecle=[]\n",
    "        self.listegca=[]\n",
    "        for i in self.blastp.keys():\n",
    "            self.listegca.append(i[0])\n",
    "            self.listegca.append(i[1])\n",
    "            self.listemenu.append(self.genomes[i[0]][0]+' vs '+self.genomes[i[1]][0])\n",
    "            self.listecle.append(self.genomes[i[0]][0])\n",
    "            self.listecle.append(self.genomes[i[1]][0])\n",
    "        \n",
    "        # Création des dossiers de sauvegarde \n",
    "        a=!dir\n",
    "        dir_graph=False\n",
    "        dir_synt=False\n",
    "        for ligne in a:\n",
    "            if '<DIR>' in ligne.split(' '):\n",
    "                if ligne.split(' ')[-1]=='Graphiques':\n",
    "                    dir_graph=True\n",
    "                if ligne.split(' ')[-1]=='Syntenies':\n",
    "                    dir_synt=True\n",
    "        \n",
    "        if dir_graph==False:\n",
    "            os.mkdir('./Graphiques')\n",
    "        if dir_synt==False:\n",
    "            os.mkdir('./Syntenies')\n",
    "        \n",
    "        \n",
    "        self.interface_para()\n",
    "        \n",
    "    def interface_para(self):\n",
    "        self.fenetre=Tk()\n",
    "        \n",
    "        self.fenetre.title('Comparaison de paires de génomes')\n",
    "        self.fenetre.geometry('1000x600')\n",
    "        \n",
    "        # Choix des organismes à comparer\n",
    "        self.description_1=Label(self.fenetre,text=\"\\nOrganismes à comparer\\n\",font=('Helvetica',20))\n",
    "        self.description_1.pack(side=TOP)\n",
    "        self.initial=StringVar(self.fenetre)\n",
    "        self.initial.set(self.listemenu[0])\n",
    "        Choixorga=OptionMenu(self.fenetre,self.initial,*self.listemenu)\n",
    "        Choixorga.config(width=90,font=('Helvetica',12))\n",
    "        Choixorga.pack(side=TOP)\n",
    "        \n",
    "        # Choix des paramètres\n",
    "        self.zonepara=Frame(self.fenetre)\n",
    "        self.description_2=Label(self.zonepara,text=\"\\nDéfinissez les paramètres pour la comparaison\\n\",font=('Helvetica',20))\n",
    "        self.description_2.pack(side=TOP)\n",
    "        \n",
    "        # Paramètre : pourcentage d'identité \n",
    "        self.pident_Frame=Frame(self.zonepara)\n",
    "            # Définition des variables\n",
    "        self.value_pident_binaire = IntVar()\n",
    "        self.value_pident_binaire.set(0)\n",
    "        \n",
    "            # Frame de gauche, pour savoir si on tient compte de ce paramètre\n",
    "        self.check_pident=Checkbutton(self.pident_Frame,text=\"Selon le pourcentage d'identité\",font=('Helvetica',12),variable=self.value_pident_binaire)\n",
    "        self.check_pident.pack(side=LEFT)\n",
    "        \n",
    "            # Frame de droite, pour déterminer un seuil\n",
    "        self.pident_seuil_Frame=Frame(self.pident_Frame)\n",
    "        self.pident_seuil_widget=Entry(self.pident_seuil_Frame)\n",
    "        self.pident_seuil_indication=Label(self.pident_seuil_Frame,text=\"          seuil (entre 0 et 1, 0.2 par défaut) =\",font=('Helvetica',12))\n",
    "        self.pident_seuil_widget.pack(side=RIGHT)\n",
    "        self.pident_seuil_indication.pack(side=LEFT)\n",
    "        self.pident_seuil_Frame.pack(side=RIGHT)\n",
    "        self.pident_Frame.pack(side=TOP)\n",
    "        \n",
    "        # Espace entre les paramètres\n",
    "        self.espace1=Label(self.zonepara,text=\" \")\n",
    "        self.espace1.pack(side=TOP)\n",
    "\n",
    "        # Paramètre : e-value\n",
    "        \n",
    "            #définition des variables\n",
    "        self.value_blastp_binaire=IntVar()\n",
    "        self.value_blastp_binaire.set(0)\n",
    "        \n",
    "        self.value_cd_binaire=IntVar()\n",
    "        self.value_cd_binaire.set(0)\n",
    "            \n",
    "            # CD-search\n",
    "        self.cd_Frame=Frame(self.zonepara)\n",
    "        self.cd_Frame_choix=Frame(self.cd_Frame)\n",
    "            # Frame de gauche, pour savoir si on tient compte de ce paramètre\n",
    "        self.check_cd=Checkbutton(self.cd_Frame_choix, variable=self.value_cd_binaire)\n",
    "        self.options=['Hit COG','Annotation fonctionnelle']\n",
    "        self.initial2=StringVar(self.cd_Frame_choix)\n",
    "        self.initial2.set(self.options[0])\n",
    "        self.Choixcd=OptionMenu(self.cd_Frame_choix,self.initial2,*self.options)\n",
    "        self.check_cd.pack(side=LEFT)\n",
    "        self.Choixcd.pack(side=RIGHT)\n",
    "        self.cd_Frame_choix.pack(side=LEFT)\n",
    "        \n",
    "            # Frame de droite, pour déterminer un seuil\n",
    "        self.cd_seuil_Frame=Frame(self.cd_Frame)\n",
    "        self.cd_seuil_widget=Entry(self.cd_seuil_Frame)\n",
    "        self.cd_seuil_indication=Label(self.cd_seuil_Frame,text=\"\\t          seuil (1e-4 par défaut) =\",font=('Helvetica',12))\n",
    "        self.cd_seuil_widget.pack(side=RIGHT)\n",
    "        self.cd_seuil_indication.pack(side=LEFT)\n",
    "        self.cd_seuil_Frame.pack(side=RIGHT)\n",
    "        self.cd_Frame.pack(side=TOP)\n",
    "        \n",
    "        # Espace entre les paramètres\n",
    "        \n",
    "        self.espace2=Label(self.zonepara,text=\" \")\n",
    "        self.espace2.pack(side=TOP)\n",
    "        \n",
    "            # Blastp\n",
    "        self.bp_Frame=Frame(self.zonepara)\n",
    "        \n",
    "            # Pour savoir si on tient compte de ce paramètre\n",
    "        self.check_bp=Checkbutton(self.bp_Frame,text=\"Selon la e-value du Blastp         \",font=('Helvetica',12),variable=self.value_blastp_binaire)\n",
    "        self.check_bp.pack(side=LEFT)\n",
    "        \n",
    "            # Frame de droite, pour déterminer un seuil\n",
    "        self.bp_seuil_Frame=Frame(self.bp_Frame)\n",
    "        self.bp_seuil_widget=Entry(self.bp_seuil_Frame)\n",
    "        self.bp_seuil_indication=Label(self.bp_seuil_Frame,text=\"\\t          seuil (1e-4 par défaut) =\",font=('Helvetica',12))\n",
    "        self.bp_seuil_widget.pack(side=RIGHT)\n",
    "        self.bp_seuil_indication.pack(side=LEFT)\n",
    "        self.bp_seuil_Frame.pack(side=RIGHT)\n",
    "        self.bp_Frame.pack(side=TOP)\n",
    "        \n",
    "        # Espace entre les paramètres      \n",
    "        self.espace3=Label(self.zonepara,text=\" \")\n",
    "        self.espace3.pack(side=TOP)\n",
    "\n",
    "        # Paramètre : couverture du hit\n",
    "        self.hit_Frame=Frame(self.zonepara)\n",
    "        self.value_hit_binaire = IntVar()\n",
    "        self.value_hit_binaire.set(0)\n",
    "        \n",
    "            # Frame de gauche, pour savoir si on tient compte de ce paramètre\n",
    "        self.check_hit=Checkbutton(self.hit_Frame,text=\"Selon la couverture du hit         \",font=('Helvetica',12),variable=self.value_hit_binaire)\n",
    "        self.check_hit.pack(side=LEFT)\n",
    "        \n",
    "            # Frame de droite, pour déterminer un seuil\n",
    "        self.hit_seuil_Frame=Frame(self.hit_Frame)\n",
    "        self.hit_seuil_widget=Entry(self.hit_seuil_Frame)\n",
    "        self.hit_seuil_indication=Label(self.hit_seuil_Frame,text=\"          seuil (entre 0 et 1, 0.2 par défaut) =\",font=('Helvetica',12))\n",
    "        self.hit_seuil_widget.pack(side=RIGHT)\n",
    "        self.hit_seuil_indication.pack(side=LEFT)\n",
    "        self.hit_seuil_Frame.pack(side=RIGHT)\n",
    "        self.hit_Frame.pack(side=TOP)\n",
    "                \n",
    "        # Espace entre les paramètres\n",
    "        self.espace4=Label(self.zonepara,text=\" \")\n",
    "        self.espace4.pack(side=TOP)\n",
    "                \n",
    "        # Bouton de fermeture du programme\n",
    "        self.bouton_close=Button(self.zonepara,text=\"CLOSE\",command=self.fenetre.destroy)\n",
    "        self.bouton_close.pack(side=BOTTOM)\n",
    "        \n",
    "        # Bouton de lancement de l'analyse\n",
    "        self.bouton_go=Button(self.zonepara,text=\"GO\",command=self.interface_dotplot)\n",
    "        self.bouton_go.pack(side=BOTTOM)\n",
    "        \n",
    "        self.zonepara.pack()\n",
    "        self.fenetre.mainloop()\n",
    "        \n",
    "        \n",
    "    def interface_dotplot(self):\n",
    "        \n",
    "        # Quels sont les organismes à comparer\n",
    "        indicechoix=self.listemenu.index(self.initial.get())\n",
    "        self.orga1,self.orga2=self.listecle[indicechoix*2],self.listecle[indicechoix*2+1]\n",
    "\n",
    "        titre=\"Résultat de la comparaison des génomes de\"+self.orga1+\"et\"+self.orga2\n",
    "        self.orga1=self.listegca[indicechoix*2]      #code GCA numéro 1\n",
    "        self.orga2=self.listegca[indicechoix*2+1]      #code GCA numéro 2       \n",
    "        #on vérifie qu'on a bien les fichiers blast et les protéomes de ces organismes\n",
    "        #pas besoin de vérifier si on a le blastp étant donné qu'on sait qu'il faut comparer ces deux organismes via le fichier blastp\n",
    "        if self.orga1 in self.proteomes.keys() and self.orga2 in self.proteomes.keys() and self.orga1 in self.cd.keys() and self.orga1 in self.cd.keys():\n",
    "            #création des listes de nom de protéines dans l'ordre\n",
    "            listeorga1=[]\n",
    "            listeorga2=[]\n",
    "            for i in range(len(self.proteomes[self.orga1])):\n",
    "                listeorga1.append(self.proteomes[self.orga1][i][0])\n",
    "            for i in range(len(self.proteomes[self.orga2])):\n",
    "                listeorga2.append(self.proteomes[self.orga2][i][0])\n",
    "    \n",
    "            coordonnees_x=[]\n",
    "            coordonnees_y=[]\n",
    "            \n",
    "            #seuils de départ\n",
    "            self.value_hit_seuil=0.2\n",
    "            self.value_blastp_seuil=1e-4\n",
    "            self.value_cd_seuil=1e-4\n",
    "            self.value_pident_seuil=0.2\n",
    "        \n",
    "            #on récupère le seuil indiqué par l'utilisateur seulement s'il l'a écrit dans le bon format avec ce motif\n",
    "            motif_seuil=re.compile('[0-9]{1}e-[0-9]+')        \n",
    "            motif_pourcentage=re.compile('0.[0-9]+')\n",
    "            # Paramètre : e-value du blastp\n",
    "            if self.value_blastp_binaire.get()==1:\n",
    "                if motif_seuil.match(self.bp_seuil_widget.get()):\n",
    "                    self.value_blastp_seuil=float(self.bp_seuil_widget.get())\n",
    "\n",
    "            #on récupère ce qui match avec ce paramètre\n",
    "                for couple in self.blastp[self.orga1,self.orga2].keys():\n",
    "                    if float(self.blastp[self.orga1,self.orga2][couple][8])<=self.value_blastp_seuil:\n",
    "                        coordonnees_x.append(listeorga1.index(couple[0]))\n",
    "                        coordonnees_y.append(listeorga2.index(couple[1]))\n",
    "        \n",
    "        \n",
    "       \n",
    "            # Paramètre : pourcentage d'identité\n",
    "            if self.value_pident_binaire.get()==1:\n",
    "                if motif_pourcentage.match(self.pident_seuil_widget.get()):\n",
    "                    self.value_pident_seuil=float(self.pident_seuil_widget.get())\n",
    "                if len(coordonnees_x)==0:\n",
    "                    for couple in self.blastp[self.orga1,self.orga2].keys():\n",
    "                        if float(self.blastp[self.orga1,self.orga2][couple][0])/100>=self.value_pident_seuil:\n",
    "                            coordonnees_x.append(listeorga1.index(couple[0]))\n",
    "                            coordonnees_y.append(listeorga2.index(couple[1]))\n",
    "                else:\n",
    "                    xtemp=[]\n",
    "                    ytemp=[]\n",
    "                    for i in range(len(coordonnees_x)):\n",
    "                        xi=coordonnees_x[i]\n",
    "                        yi=coordonnees_y[i]\n",
    "                        compare_pourcentage=float(self.blastp[self.orga1,self.orga2][listeorga1[xi],listeorga2[yi]][0])\n",
    "                        if compare_pourcentage/100>=self.value_pident_seuil:\n",
    "                            xtemp.append(coordonnees_x[i])\n",
    "                            ytemp.append(coordonnees_y[i])\n",
    "                    coordonnees_x=copy.deepcopy(xtemp)\n",
    "                    coordonnees_y=copy.deepcopy(ytemp)\n",
    "            self.fenetre2=Tk()\n",
    "            self.fenetre2.geometry('1000x600')\n",
    "            self.fenetre2.title(self.initial.get())\n",
    "        \n",
    "            # Paramètre : couverture du hit\n",
    "            if self.value_hit_binaire.get()==1:\n",
    "                if motif_pourcentage.match(self.hit_seuil_widget.get()):\n",
    "                    self.value_hit_seuil=float(self.hit_seuil_widget.get())\n",
    "                #Dans le cas où c'est le premier paramètre pris en compte\n",
    "                if len(coordonnees_x)==0:\n",
    "                    for couple in self.blastp[self.orga1,self.orga2].keys():\n",
    "                        pident=float(self.blastp[self.orga1,self.orga2][couple][0])/100\n",
    "                        longueur=float(self.blastp[self.orga1,self.orga2][couple][1])\n",
    "                        nb_mismatch=float(self.blastp[self.orga1,self.orga2][couple][2])\n",
    "                        nb_gap=(1-pident)*longueur-nb_mismatch  #le pourcentage d'identité est arrondi au centième près, donc\n",
    "                                                                #le nombre de gap n'est pas un entier, donc il sera arrondi par la suite\n",
    "                        longueur_prot=self.proteomes[self.orga1][listeorga1.index(couple[0])][1]\n",
    "                        couverture=(longueur-nb_gap)/longueur_prot\n",
    "                        if couverture>=self.value_hit_seuil:\n",
    "                            coordonnees_x.append(listeorga1.index(couple[0]))\n",
    "                            coordonnees_y.append(listeorga2.index(couple[1]))\n",
    "                else:\n",
    "                    xtemp=[]\n",
    "                    ytemp=[]\n",
    "                    for i in range(len(coordonnees_x)):\n",
    "                        xi=coordonnees_x[i]\n",
    "                        yi=coordonnees_y[i]\n",
    "                        pident=float(self.blastp[self.orga1,self.orga2][listeorga1[xi],listeorga2[yi]][0])/100\n",
    "                        longueur=float(self.blastp[self.orga1,self.orga2][listeorga1[xi],listeorga2[yi]][1])\n",
    "                        nb_mismatch=float(self.blastp[self.orga1,self.orga2][listeorga1[xi],listeorga2[yi]][2])\n",
    "                        nb_gap=(1-pident)*longueur-nb_mismatch  #le pourcentage d'identité est arrondi au centième près, donc\n",
    "                                                                #le nombre de gap n'est pas un entier, donc il sera arrondi par la suite\n",
    "                        longueur_prot=self.proteomes[self.orga1][listeorga1.index(listeorga1[xi])][1]\n",
    "                        couverture=(longueur-nb_gap)/longueur_prot\n",
    "                        if couverture>=self.value_hit_seuil:\n",
    "                            xtemp.append(listeorga1.index(listeorga1[xi]))\n",
    "                            ytemp.append(listeorga2.index(listeorga2[yi]))\n",
    "                    coordonnees_x=copy.deepcopy(xtemp)\n",
    "                    coordonnees_y=copy.deepcopy(ytemp)\n",
    "\n",
    "               # Paramètre : cd-search\n",
    "            if self.value_cd_binaire.get()==1:\n",
    "                if motif_seuil.match(self.cd_seuil_widget.get()):\n",
    "                    self.value_cd_seuil=float(self.cd_seuil_widget.get())\n",
    "                indicechoixcd=self.options.index(self.initial2.get())\n",
    "                #Dans le cas où c'est le premier paramètre pris en compte\n",
    "                #COG\n",
    "                if indicechoixcd==0:\n",
    "                    if len(coordonnees_x)==0:\n",
    "                        for i in self.cd[self.orga1].keys():\n",
    "                            for j in self.cd[self.orga2].keys():\n",
    "                                if float(self.cd[self.orga1][i][2])<=self.value_cd_seuil and float(self.cd[self.orga2][j][2])<=self.value_cd_seuil:\n",
    "                                    if self.cd[self.orga1][i][4]==self.cd[self.orga2][j][4]:\n",
    "                                        coordonnees_x.append(listeorga1.index(i))\n",
    "                                        coordonnees_y.append(listeorga2.index(j))\n",
    "                    #Dans le cas où ça n'est pas le premier paramètre pris en compte\n",
    "                    else:\n",
    "                        xtemp=[]\n",
    "                        ytemp=[]\n",
    "                        for i in range(len(coordonnees_x)):\n",
    "                            prot1=listeorga1[coordonnees_x[i]]\n",
    "                            prot2=listeorga2[coordonnees_y[i]]\n",
    "                            if prot1 in self.cd[self.orga1].keys() and prot2 in self.cd[self.orga2].keys():\n",
    "                                if float(self.cd[self.orga1][prot1][2])<=self.value_cd_seuil and float(self.cd[self.orga2][prot2][2])<=self.value_cd_seuil:\n",
    "                                    if self.cd[self.orga1][prot1][4]==self.cd[self.orga2][prot2][4]:\n",
    "                                        xtemp.append(coordonnees_x[i])\n",
    "                                        ytemp.append(coordonnees_y[i])\n",
    "                        coordonnees_x=copy.deepcopy(xtemp)\n",
    "                        coordonnees_y=copy.deepcopy(ytemp)\n",
    "            \n",
    "            \n",
    "            \n",
    "                elif indicechoixcd==1:\n",
    "                    #annotation fonctionnelle\n",
    "                    #Dans le cas où c'est le premier paramètre pris en compte\n",
    "                    if len(coordonnees_x)==0:\n",
    "                        for i in self.cd[self.orga1].keys():\n",
    "                            for j in self.cd[self.orga2].keys():\n",
    "                                if float(self.cd[self.orga1][i][2])<=self.value_cd_seuil and float(self.cd[self.orga2][j][2])<=self.value_cd_seuil:\n",
    "                                    if self.cognames[self.cd[self.orga1][i][4]][0]==self.cognames[self.cd[self.orga2][j][4]][0]:\n",
    "                                        coordonnees_x.append(listeorga1.index(i))\n",
    "                                        coordonnees_y.append(listeorga2.index(j))\n",
    "                    #Dans le cas où ça n'est pas le premier paramètre pris en compte\n",
    "                    else:\n",
    "                        xtemp=[]\n",
    "                        ytemp=[]\n",
    "                        for i in range(len(coordonnees_x)):\n",
    "                            prot1=listeorga1[coordonnees_x[i]]\n",
    "                            prot2=listeorga2[coordonnees_y[i]]\n",
    "                            if prot1 in self.cd[self.orga1].keys() and prot2 in self.cd[self.orga2].keys():\n",
    "                                if float(self.cd[self.orga1][prot1][2])<=self.value_cd_seuil and float(self.cd[self.orga2][prot2][2])<=self.value_cd_seuil:\n",
    "                                        if self.cognames[self.cd[self.orga1][prot1][4]][0]==self.cognames[self.cd[self.orga2][prot2][4]][0]:\n",
    "                                            xtemp.append(coordonnees_x[i])\n",
    "                                            ytemp.append(coordonnees_y[i])\n",
    "                        coordonnees_x=copy.deepcopy(xtemp)\n",
    "                        coordonnees_y=copy.deepcopy(ytemp)\n",
    "            \n",
    "            #On indique sur la fenêtre du graphique les paramètres pris en compte\n",
    "            paras=\"Paramètres pris en compte :\"\n",
    "            Aucun=True\n",
    "            if self.value_blastp_binaire.get()==1:\n",
    "                paras+=\"\\nLa e-value associée au blastp avec un seuil de \"+str(self.value_blastp_seuil)\n",
    "                Aucun=False\n",
    "            if self.value_pident_binaire.get()==1:\n",
    "                paras+=\"\\nLe pourcentage d'identité avec un seuil de \"+str(self.value_pident_seuil*100)+\"%\"\n",
    "                Aucun=False\n",
    "            if self.value_hit_binaire.get()==1:\n",
    "                paras+=\"\\nLa couverture du hit avec un seuil de \"+str(self.value_hit_seuil*100)+\"%\"\n",
    "                Aucun=False\n",
    "            if self.value_cd_binaire.get()==1:\n",
    "                if indicechoixcd==0:\n",
    "                    paras+=\"\\nLa e-value associée au COG avec un seuil de \"+str(self.value_cd_seuil)\n",
    "                if indicechoixcd==1:\n",
    "                    paras+=\"\\nL'annotation fonctionnelle avec un seuil de \"+str(self.value_cd_seuil)\n",
    "                Aucun=False\n",
    "            if Aucun==True:\n",
    "                paras+=\"\\nAucun\"\n",
    "            paras+=\"\\n\"\n",
    "             \n",
    "            label_para=Label(self.fenetre2,text=paras,font=('Helvetica',12))    \n",
    "            label_para.pack(side=TOP)\n",
    "        \n",
    "            fig=Figure(figsize=(6,4),dpi=96)\n",
    "            ax=fig.add_subplot(111)\n",
    "            ax.scatter(coordonnees_x,coordonnees_y,s=0.1)\n",
    "        \n",
    "            self.bouton_close=Button(self.fenetre2,text=\"CLOSE\",command=self.fenetre2.destroy)\n",
    "            self.bouton_close.pack(side=BOTTOM)\n",
    "\n",
    "            \n",
    "            name=self.orga1+\"vs\"+self.orga2           \n",
    "            \n",
    "            graph=FigureCanvasTkAgg(fig, master=self.fenetre2)\n",
    "            graph.get_default_filename = lambda: name\n",
    "            graph.draw()\n",
    "            graph.get_tk_widget().pack(side=TOP, fill=BOTH, expand=1)\n",
    "\n",
    "            toolbar = NavigationToolbar2Tk(graph, self.fenetre2)\n",
    "            toolbar.update()\n",
    "            graph.get_tk_widget().pack(side=TOP, fill=BOTH, expand=1)\n",
    "        \n",
    "            def on_key_press(event):\n",
    "                print(\"you pressed {}\".format(event.key))\n",
    "                key_press_handler(event, canvas, toolbar)\n",
    "\n",
    "\n",
    "            graph.mpl_connect(\"key_press_event\", on_key_press)\n",
    "        \n",
    "        \n",
    "        \n",
    "            self.fenetre2.mainloop()\n",
    "       \n",
    "        else:\n",
    "            if self.orga1 not in self.proteomes.keys():\n",
    "                print(\"Il manque le fichier protéome de \",self.orga1)\n",
    "            if self.orga2 not in self.proteomes.keys():\n",
    "                print(\"Il manque le fichier protéome de \",self.orga2)\n",
    "            if self.orga1 not in self.cd.keys():\n",
    "                print(\"Il manque le fichier CD-search de \",self.orga1)\n",
    "            if self.orga1 not in self.cd.keys():\n",
    "                print(\"Il manque le fichier CD-search de \",self.orga2)\n",
    "\n",
    "\n",
    "        \n",
    "a=interface()\n",
    "\n",
    "\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "> rename des graph en fonction des para ????\n",
    "> analyse blocs de synténie"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
